src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function getindex(a::Hecke.SRow, b::AbstractArray{Int, 1})
+  if length(a.pos) == 0
+    return a
+  end
+  m = minimum(b)
+  b = sparse_row(parent(a.values[1]))
+  for (k,v) = a
+    if k in b
+      push!(b.pos, k-b+1)
+      push!(b.values, v)
+    end
+  end
+  return b
+end
+
+@doc Markdown.doc"""
+    sparse_row(A::MatElem)
+
+Convert `A` to a sparse row. 
+`nrows(A) == 1` must hold.
+"""
+function sparse_row(A::MatElem)
+  @assert nrows(A) == 1
+  return Hecke.sparse_matrix(A)[1]
+end
+
+@doc Markdown.doc"""
+    dense_row(r::Hecke.SRow, n::Int)
+
+Convert `r[1:n]` to a dense row, that is an AbstractAlgebra matrix.
+"""
+function dense_row(r::Hecke.SRow, n::Int)
+  R = base_ring(r)
+  A = zero_matrix(R, 1, n)
+  for i in intersect(r.pos, 1:n)
+    A[1,i] = r[i]
+  end
+  return A
+end
+
+##############################
+#should be in Singular.jl
+function Singular.intersection(a::Singular.smodule, b::Singular.smodule)
+  c = base_ring(a)
+  return Singular.Module(c, Singular.libSingular.id_Intersection(a.ptr, b.ptr, c.ptr))
+end
+
+function _reduce(a::Singular.smodule, b::Singular.smodule)
+  @assert b.isGB
+  p = Singular.libSingular.p_Reduce(a.ptr, b.ptr, base_ring(b).ptr)
+  return Singular.Module(base_ring(b), p)
+end
+
+function _reduce(a::Singular.svector, b::Singular.smodule)
+  @assert b.isGB
+  p = _reduce(Singular.Module(base_ring(b), a), b)[1]
+  return Singular.Module(base_ring(b), p)[1]
+end
+
+#TODO: tensor_product from Raul's H is broken
+
+
+######################################
+# Migrating test
+######################################
+@doc Markdown.doc"""
+    projection(F::FreeMod, indices::AbstractArray)
+
+Return the canonical projection from $F = R^I$ to $R^(\texttt{indices})$ where $\texttt{indices} \subset I$.
+"""
+function projection(F::FreeMod, indices::AbstractArray)
+  @assert all(x -> x <= ngens(F), indices)
+  @assert length(Set(indices)) == length(indices) # unique indices
+  R = base_ring(F)
+  G = FreeMod(R, length(indices))
+  return hom(F, G, [i in indices ? G[findfirst(x->x==i,indices)] : zero(G) for i=1:ngens(F)])
+end
+
+@doc Markdown.doc"""
+    preimage(H::SubQuoHom,elems::Vector{SubQuoElem{T}}, task::Symbol = :none) where {T}
+
+Return the preimage of the submodule generated by the Elements `elems` under $H$
+as a subquotient, as well as the injection homomorphism into the domain of $H$.
+"""
+function preimage(H::SubQuoHom,elems::Vector{SubQuoElem{T}}, task::Symbol = :none) where {T}
+  if length(elems)==0
+      throw(ArgumentError("too few arguments"))
+  end
+  R = base_ring(domain(H))
+  row_length = ngens(codomain(H))
+  submod = vcat((dense_row(coeffs(e), row_length) for e in elems)...)
+  C = matrix(present_as_cokernel(codomain(H)).quo)
+  A = vcat(matrix(H), C, submod)
+  G = FreeMod(R, nrows(A))
+  A = FreeModuleHom(G, FreeMod(R, ncols(A)), A)
```

Comment:
I don't really like that so much code in this PR works again with raw matrix manipulation. For me, a big part of the appeal of Oscar is that we can get away from that and do things in a higher level. 

Like, here, I'd expect that one would first compute preimages of the elements in `elems`, then compute the submodule spanned by these elements plus the kernel of `H`.